service/controller/v12/key/key.go

```diff
@@ -300,6 +300,11 @@ func RouteTableName(customObject providerv1alpha1.AzureConfig) string {
 	return fmt.Sprintf("%s-%s", ClusterID(customObject), routeTableSuffix)
 }
 
+// AvailabilityZones returns the availability zones where the cluster will be created.
+func AvailabilityZones(customObject providerv1alpha1.AzureConfig) []int {
+	return customObject.Spec.Azure.AvailabilityZones
```

Comment:
Might the key be missing from the CR? What happens in that case?

Reply:
It will break on that case I think.

I'm not sure I understand how versioning works. Since we didn't flatten our operator, our `azure-operator` reconciliates different versions of the `AzureConfig` CRD, depending on the `VersionBundle` field. So if my understanding is correct, old CR's without the `AvailabilityZones` field will be handled by old versions of the `azure-operator` , while new CR's will be handled by this new version. That way nothing breaks. Does that make sense?

Reply:
There's actually more to it:
* When something with latest `apiextensions` updates the CR, the field gets added with empty value as it's not `omitempty`: https://github.com/giantswarm/apiextensions/pull/300/files#diff-f8ea2fc1fb00586a596f9bc54c5ebbb5R68
* No matter whether or not the field exists in CR, the data structure here is not using pointers so it's just empty list.
* In general Go handles nil slices gracefully so empty value should work.
* With the release that introduces multi-AZ, it must have migration code that updates CR with present AZ configuration.